### PR TITLE
Add const generics support for pattern types

### PR DESCRIPTION
**Pull request by spastorino**
r? @oli-obk

### CODE REVIEW

File: compiler/rustc_middle/src/ty/layout.rs [hunk 1]
```diff
@@ -312,6 +312,11 @@ impl<'tcx> SizeSkeleton<'tcx> {
     ) -> Result<SizeSkeleton<'tcx>, &'tcx LayoutError<'tcx>> {
         debug_assert!(!ty.has_non_region_infer());
 
+        let ty = match *ty.kind() {
+            ty::Pat(inner_ty, pat) if matches!(*pat, ty::PatternKind::Range { .. }) => inner_ty,
```
Comment (oli-obk) [inline compiler/rustc_middle/src/ty/layout.rs:316]: Why is this needed?

Comment (spastorino) [reply compiler/rustc_middle/src/ty/layout.rs]: It is not for this PR, I think this was needed to make newindex use pattern types or some of the other conversion I was doing from rustc_scalar attributes to pattern types but this changed ended sneaking into this PR.

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs [hunk 2]
```diff
@@ -2223,6 +2223,21 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
...
                                     }
                                 }
+
+                                hir::ExprKind::Path(hir::QPath::Resolved(
+                                    _,
+                                    &hir::Path {
+                                        res: Res::Def(DefKind::ConstParam, def_id), ..
+                                    },
+                                )) => {
+                                    let ty = tcx.type_of(def_id).instantiate_identity();
```
Comment (fmease) [inline compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:2233]:
```suggestion
                                    let ty = tcx.type_of(def_id).no_bound_vars().expect("const parameter types cannot be generic");
```

**Review by fmease — CHANGES_REQUESTED** [review]

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs (hunk 2 above)
Comment (fmease) [reply compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:2233]: Ideally we would factor out the logic common between this here and `lower_const_param` 🤔, not blocking through

Comment (oli-obk) [reply compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:2233]: Slowly wondering if we should just use entirely new hir nodes so we can just use AnonConst like other type system consts

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs [hunk 3]
```diff
@@ -2223,6 +2223,24 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
                                         Err(LitToConstError::TypeError) => todo!(),
                                     }
                                 }
+
+                                hir::ExprKind::Path(hir::QPath::Resolved(
+                                    _,
+                                    &hir::Path {
+                                        res: Res::Def(DefKind::ConstParam, def_id), ..
+                                    },
+                                )) => {
```
Comment (fmease) [inline compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:2232]: Note that we're missing a call to `prohibit_generic_args` here. Right now, we permit:

```rs
pattern_type!(u32 is START::<(), i32, 2>..=END::<_>)
```

While we could add this check, this just reinforces oli's point: https://github.com/rust-lang/rust/pull/123689#discussion_r1558148361, [Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/pattern.20type.20HIR.20nodes/near/432410768).

Comment (fmease) [reply compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:2232]: Just noticed this while reviewing an unrelated PR.
